compiler-plugin/src/main/java/io/ballerina/stdlib/persist/sql/compiler/codemodifier/PersistQueryValidator.java

```diff
@@ -0,0 +1,282 @@
+                                 Map<String, String> variables, Map<QueryPipelineNode, Query> queries,
+                                 Map<QueryPipelineNode, Query> validatedQueries,
+                                 List<String> persistClientVariableNames) {
+        this.entities = entities;
+        this.persistClientNames = persistClientNames;
+        this.variables = variables;
+        this.queries = queries;
+        this.validatedQueries = validatedQueries;
+        this.persistClientVariableNames = persistClientVariableNames;
+    }
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext ctx) {
+        if (Utils.hasCompilationErrors(ctx)) {
+            return;
+        }
+
+        QueryPipelineNode queryPipelineNode = (QueryPipelineNode) ctx.node();
+        FromClauseNode fromClauseNode = queryPipelineNode.fromClause();
+        Query query = isQueryUsingPersistentClient(fromClauseNode);
+        if (query == null) {
+            return;
+        }
+
+        NodeList<IntermediateClauseNode> intermediateClauseNodes = queryPipelineNode.intermediateClauses();
+        List<IntermediateClauseNode> whereClauseNodes = intermediateClauseNodes.stream()
+                .filter((node) -> node instanceof WhereClauseNode)
+                .collect(Collectors.toList());
+
+        List<IntermediateClauseNode> orderByClauseNode = intermediateClauseNodes.stream()
+                .filter((node) -> node instanceof OrderByClauseNode)
+                .collect(Collectors.toList());
+
+        List<IntermediateClauseNode> limitClauseNode = intermediateClauseNodes.stream()
+                .filter((node) -> node instanceof LimitClauseNode)
+                .collect(Collectors.toList());
+
+        List<IntermediateClauseNode> groupByClauseNode = intermediateClauseNodes.stream()
+                .filter((node) -> node instanceof GroupByClauseNode)
+                .collect(Collectors.toList());
+
+        boolean isWhereClauseUsed = whereClauseNodes.size() != 0;
+        boolean isOrderByClauseUsed = orderByClauseNode.size() != 0;
+        boolean isLimitClauseUsed = limitClauseNode.size() != 0;
+        boolean groupByClauseUsed = groupByClauseNode.size() != 0;
+
+        if (!isWhereClauseUsed && !isOrderByClauseUsed && !isLimitClauseUsed && !groupByClauseUsed) {
+            return;
+        }
+
+        List<IntermediateClauseNode> letClauseNodes = intermediateClauseNodes.stream()
+                .filter((node) -> node.kind().equals(SyntaxKind.LET_CLAUSE))
+                .collect(Collectors.toList());
+
+        if (groupByClauseUsed) {
+            NodeLocation groupByClauseLocation = groupByClauseNode.get(0).location();
+            int groupByClauseEndLine = groupByClauseLocation.lineRange().startLine().line();
+            if (isWhereClauseUsed) {
+                if (groupByClauseEndLine < whereClauseNodes.get(0).location().lineRange().startLine().line()) {
```

Comment:
```suggestion
            if (isWhereClauseUsed) {
            	int whereClauseStartLine = whereClauseNodes.get(0).location().lineRange().startLine().line();
                if (groupByClauseEndLine < whereClauseStartLine) {
```

Might make the code more readable